### PR TITLE
Fix UsedNamespaceOrType::Equals

### LINKED ISSUE #7015: UsedNamespaceOrType comparing type argument incorrectly
Consider the contents of the `Equals` method

``` csharp
return AliasOpt == other.AliasOpt
    && TargetAssemblyOpt == other.TargetAssemblyOpt
    && TargetNamespaceOpt == other.TargetNamespaceOpt
    && TargetTypeOpt == other.TargetTypeOpt
    && TargetXmlNamespaceOpt == other.TargetXmlNamespaceOpt;
```

This is comparing `ITypeReference` values by reference instead of using their `Equals` method.  In the case of generic types the `ITypeReference` values can be different objects for equivalent symbols and this comparison will incorrectly return false.  

This is actually breaking determinism on compilations that have a using alias where the target type is a bound generic.  It causes the result of `ShouldForwardNamespaceScopes` to be non-deterministic and hence changes the PDB output and content id hash.

### PR DESCRIPTION
**Pull request by jaredpar**
The Equals method was using reference equality to compare ITypeReference instances.  Correct comparison of that type must use .Equals as equivalent generic values can be different objects.

This surfaced as a bug in determinism.  In the case we had an alias'd using where the target type was generic this Equals method came into play.  Based on whether we got a generic cache hit or not ended up changing the PDB output and as an effect the determinism id of the resulting assembly.

closes #7015

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

**Comment by davkean** [comment]
Looks like real failures: 

```
error CS2001: Source file '/mnt/j/workspace/roslyn_prtest_lin_dbg_unit32/src/Compilers/Core/CodeAnalysisTest/PEWriter/UsedNamespaceOrTypeTests.cs' could not be found. [/mnt/j/workspace/roslyn_prtest_lin_dbg_unit32/src/Compilers/Core/CodeAnalysisTest/CodeAnalysisTest.csproj]
```

**Comment by jaredpar** [comment]
@davkean think it's a file path casing issue. Will fix.

**Comment by TyOverby** [comment]
:+1:  Good catch!

File: src/Compilers/Core/Portable/PEWriter/UsedNamespaceOrType.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ public bool Equals(UsedNamespaceOrType other)
             return AliasOpt == other.AliasOpt
                 && TargetAssemblyOpt == other.TargetAssemblyOpt
                 && TargetNamespaceOpt == other.TargetNamespaceOpt
-                && TargetTypeOpt == other.TargetTypeOpt
+                && object.Equals(TargetTypeOpt, other.TargetTypeOpt)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/UsedNamespaceOrType.cs]: Is there a reason to not do the same for other interfaces involved?

File: src/Compilers/Core/Portable/CodeAnalysis.csproj [hunk 2]
```diff
@@ -747,6 +747,7 @@
     <InternalsVisibleToTest Include="Roslyn.Test.Utilities.FX45" />
     <InternalsVisibleToTest Include="Roslyn.Test.Utilities" />
     <InternalsVisibleToTest Include="Roslyn.Compilers.CompilerServer.UnitTests" />
+    <InternalsVisibleToMoq Include="DynamicProxyGenAssembly2" />
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: Is this change required? Could you elaborate why?

**Comment by AlekseyTs** [comment]
Are changes to various .json files related to the fix?

File: src/Compilers/Core/Portable/PEWriter/UsedNamespaceOrType.cs (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/PEWriter/UsedNamespaceOrType.cs]: Is that the correct way to compare them?  If so happy to change it.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/UsedNamespaceOrType.cs]: It feels like value equality is what we want here.

**Comment by jaredpar** [comment]
@AlekseyTs yes the JSON files are a part of the fix.  I needed a reference to Moq for the unit test.

File: src/Compilers/Core/Portable/CodeAnalysis.csproj (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/CodeAnalysis.csproj]: The Moq libraries needs internals visible to in order to for `Mock<T>` declarations to function when `T` is an internal type.

**Pull request merged**